File: Leetcode/Algorithm/cpp/00121-Best Time to Buy and Sell Stock.cc
const int INF = 1 << 29;

class Solution {
public:
    int maxProfit(vector<int> &prices) {
        vector<int> leftmin, rightmax;
        int n = prices.size();
        for (int i = 0, j = INF; i < n; i++) {
            if (prices[i] < j) {
                j = prices[i];
            }
            leftmin.push_back(j);
        }
        for (int i = n - 1, j = -INF; i >= 0; i--) {
            if (prices[i] > j) {
                j = prices[i];
            }
            rightmax.push_back(j);
        }
        reverse(rightmax.begin(), rightmax.end());
        int ans = 0;
        for (int i = 0; i < n; i++) {
            ans = max(ans, rightmax[i] - leftmin[i]);
        }
        return ans;
    }
};

File: Leetcode/Algorithm/python/0000/00121-Best Time to Buy and Sell Stock.py
INF = 0xdeadbeef

class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit(self, prices):
        n = len(prices)
        if n < 2:
            return 0
        ans = 0
        maxi, mini = -INF, INF
        maxis, minis = [], []
        for i in xrange(n):
            mini = min(mini, prices[i])
            minis.append(mini)
        for i in xrange(n - 1, -1, -1):
            maxi = max(maxi, prices[i])
            maxis.append(maxi)
        maxis = maxis[::-1]
        for (a, b) in zip(minis, maxis):
            ans = max(ans, b - a)
        return ans
